kotlin/src/main/java/org/apache/zeppelin/kotlin/reflect/ContextUpdater.java

```diff
@@ -0,0 +1,150 @@
+      refreshVariables(lines);
+      refreshMethods(lines);
+    } catch (ReflectiveOperationException | NullPointerException e) {
+      logger.error("Exception updating current variables", e);
+    }
+  }
+
+  private void refreshMethods(List<Object> lines) {
+    functions.clear();
+    for (Object line : lines) {
+      Method[] methods = line.getClass().getMethods();
+      for (Method method : methods) {
+        if (objectMethods.contains(method) || method.getName().equals("main")) {
+          continue;
+        }
+        KFunction<?> function = ReflectJvmMapping.getKotlinFunction(method);
+        if (function == null) {
+          continue;
+        }
+        functions.add(new KotlinFunctionInfo(function));
+      }
+    }
+  }
+
+  private List<Object> getLines() {
+    List<Object> lines = state.getHistory().stream()
+        .map(this::getLineFromRecord)
+        .collect(Collectors.toList());
+
+    Collections.reverse(lines);
+    return lines;
+  }
+
+  private Object getLineFromRecord(ReplHistoryRecord<? extends Pair<?, ?>> record) {
+    Object statePair = record.getItem().getSecond();
+    return ((Pair<?, ?>) statePair).getSecond();
+  }
+
+  private Object getImplicitReceiver(Object script)
+      throws ReflectiveOperationException {
+    Field receiverField = script.getClass().getDeclaredField("$$implicitReceiver0");
+    return receiverField.get(script);
+  }
+
+  private void refreshVariables(List<Object> lines) throws ReflectiveOperationException {
+    vars.clear();
+    if (!lines.isEmpty()) {
+      Object receiver = getImplicitReceiver(lines.get(0));
+      findReceiverVariables(receiver);
+    }
+    for (Object line : lines) {
+      findLineVariables(line);
+    }
+  }
+
+  // For lines, we only want fields from top level class
+  private void findLineVariables(Object line) throws IllegalAccessException {
+    Field[] fields = line.getClass().getDeclaredFields();
+    findVariables(fields, line);
+  }
+
+  // For implicit receiver, we want to also get fields in parent classes
+  private void findReceiverVariables(Object receiver) throws IllegalAccessException {
+    List<Field> fieldsList = new ArrayList<>();
+    for (Class<?> cl = receiver.getClass(); cl != null; cl = cl.getSuperclass()) {
+      fieldsList.addAll(Arrays.asList(cl.getDeclaredFields()));
+    }
+    findVariables(fieldsList.toArray(new Field[0]), receiver);
+  }
+
+  private void findVariables(Field[] fields, Object o) throws IllegalAccessException {
+    for (Field field : fields) {
+      String fieldName = field.getName();
+      if (fieldName.contains("$$implicitReceiver")) {
+        continue;
+      }
+
+      field.setAccessible(true);
+      Object value = field.get(o);
+      if (!fieldName.contains("script$")) {
+        KProperty<?> descriptor = ReflectJvmMapping.getKotlinProperty(field);
+        if (descriptor != null) {
+          vars.putIfAbsent(fieldName, new KotlinVariableInfo(value, descriptor));
```

Comment:
Should we really use `putIfAbsent` here, not `put`? May it happen that field is redeclared, are there any tests?

Reply:
Updating a "variable" in REPL actually creates a field with the same name in a new Line_ class.
With the way we navigate lines (from newest to oldest) we shouldn't `put` existing fields as they would contain old values. 
This is covered with `testVariables()` test in `KotlinInterpreterTest`.